### ISSUE TITLE
imgsz adjustment

### ISSUE THREAD

**Issue by AlbertoPad** [ultralytics/ultralytics#11860]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

First of all, I would like to thank you for the wonderful work you are doing. My question comes from the following: I want to train a model with 1280x720 images, but when I change the imgsz parameter, the model input sets it to 1280x1280. I can see in the “train_batch0,train_batch1,train_batch2” images that the images are padded to 1280x1280
IMAGE
![image](https://github.com/ultralytics/ultralytics/assets/116552864/baef7128-e0fa-4d66-839d-359486bf5ce8)

CODE
`
        model = YOLO(pretrainedModel)
        DAParameters={key: dataJsonDict[key] for key in augmentationParameter}
        trainingParameters={'data': pathToYOLOYaml, 'epochs': epochs, 'imgsz': imgsz, 'batch':batch, 'save_period': 10,
        'patience': patience, 'project': savePath, **DAParameters}
        
        model.train(**trainingParameters)
`
dataJsonDict
{
    "lr0": 0.01,
    "lrf": 0.01,
    "momentum": 0.937,
    "weight_decay": 0.0005,
    "warmup_epochs": 3.0,
    "warmup_momentum": 0.8,
    "warmup_bias_lr": 0.1,
    "box": 7.5,
    "cls": 0.5,
    "dfl": 1.5,
    "pose": 0.0,
    "kobj": 1.0,
    "label_smoothing": 0.0,
    "nbs": 64,
    "hsv_h": 0.0,
    "hsv_s": 0.0,
    "hsv_v": 0.0,
    "degrees": 0.0,
    "translate": 0.0,
    "scale": 0.0,
    "shear": 0.0,
    "perspective": 0.0,
    "flipud": 0.0,
    "fliplr": 0.0,
    "mosaic": 0.0,
    "mixup": 0.0,
    "copy_paste": 0.0,
    "imgsz":[720,1280]
  }

OUTPUT TRAIN
`
      Params      GFLOPs  GPU_mem (GB)  forward (ms) backward (ms)                   input                  output
     3015528       32.88         0.700          27.4            34      (1, 3, 1280, 1280)                    list
     3015528       65.75         1.139         16.24          33.9      (2, 3, 1280, 1280)                    list
`

**Comment by glenn-jocher** [comment]
Hello! Thank you for your kind words and for reaching out with your question.

It appears that YOLOv8 by default resizes and pads the images to square dimensions to maintain consistency across the dataset, which is why you're seeing 1280x1280. This is intended to optimize processing and ensure that models behave predictably across various input sizes.

If you're looking to maintain the original aspect ratio without padding, you might need to adjust your training setup or preprocess your images to fit the model's expected input shape without padding. Adjusting the `rect` training argument to `True` in your training configuration might help as it enables rectangular training, which might preserve aspect ratios better by using minimal padding strategies:

Here's how you can adjust your code:
```python
trainingParameters = {
    'data': pathToYOLOYaml,
    'epochs': epochs,
    'imgsz': imgsz,
    'batch': batch,
    'save_period': 10,
    'patience': patience,
    'project': savePath,
    'rect': True,  # Enable rectangular training
    **DAParameters
}
```

Give this a try and let me know if it helps or if you have more questions! 😊

**Comment by AlbertoPad** [comment]
I have the same problem, Yolov8 with "rect=True" continue padding to 1280x1280.
PARAMETERS MODEL
task=detect, mode=train, model=yolov8n.pt, data=/media/alberto_workspace/DATASETS/DS_YOLO/BAR_20240215/yolo.yaml, epochs=50, time=None, patience=10, batch=-1, imgsz=[720, 1280], save=True, save_period=5, cache=False, device=None, workers=8, project=/media/alberto_workspace/MODEL_ULTRA/BAR_20240215_IMGSZ_AUG, name=train, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=True, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=0.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.0, hsv_s=0.0, hsv_v=0.0, degrees=0.0, translate=0.0, scale=0.0, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.0, bgr=0.0, mosaic=0.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=/media/alberto_workspace/MODEL_ULTRA/BAR_20240215_IMGSZ_AUG/train
Overriding model.yaml nc=80 with nc=23

OUTPUT MODEL
WARNING :warning: updating to 'imgsz=1280'. 'train' and 'val' imgsz must be an integer, while 'predict' and 'export' imgsz may be a [h, w] list or an integer, i.e. 'yolo export imgsz=640,480' or 'yolo export imgsz=640'
AutoBatch: Computing optimal batch size for imgsz=1280
AutoBatch: CUDA:0 (NVIDIA RTX A5000) 23.68G total, 0.22G reserved, 0.05G allocated, 23.40G free
      Params      GFLOPs  GPU_mem (GB)  forward (ms) backward (ms)                   input                  output
     3015333       32.87         0.700         27.52         32.64      (1, 3, 1280, 1280)                    list
     3015333       65.74         1.130         15.86         33.82      (2, 3, 1280, 1280)                    list
     3015333       131.5         2.053         15.66         37.07      (4, 3, 1280, 1280)                    list
     3015333         263         3.844         24.76         48.16      (8, 3, 1280, 1280)                    list
     3015333       525.9         7.785         48.14         74.43     (16, 3, 1280, 1280)                    list

**Comment by glenn-jocher** [comment]
Hello! 👋

It looks like the rectangular training mode (`rect=True`) may not be behaving as expected. The `imgsz` parameter should indeed allow non-square image sizes, but it seems that the training is still using a square input.

One potential issue could be related to how the `imgsz` parameter is formatted or interpreted in the code. It's important to ensure that the image size parameter conforms to the expected input structure. As the warning message indicates, training requires a single integer `imgsz`, which means it automatically pads and resizes any input to be square.

For your specific requirement (using non-square dimensions directly for training), here are a couple of strategies:

1. **Pre-process your images**: Resize or pad them to your target aspect ratio before feeding them into the model.
2. **Modify the data loading mechanism**: Adjust the data loader to handle non-square resizing and padding according to the aspect ratio you need.

Keep in mind that modifying the core functionalities can be complex and might require thorough testing to ensure the integrity of the model's performance.

I hope this points you in the right direction! Let me know if you have more questions.
